### PR TITLE
LLM: Convert draft_model kv_cache from bf16 to fp32

### PR DESCRIPTION
**Pull request by hzjane**
## Description
To solve the issue of draft_model slower, maybe We have three ways to test.
1. Convert draft_model kv_cache from bf16 to fp32 (every iter, when draft_step== 0, draft_model generate time will become  slower causing the average draft_time more slower)
2. (This way) Convert the same kv_cache from bf16 to fp32(only first init), and then the target_model will use fp32 and generate fp32 cache. (The speed of qwen's target_model has not changed, but llama has become slower)
3. Draft_model maintains kv_cache itself, and append new matched kv_cache (Maybe the accept rate will become lower，need to test)
On way1 and way3, the draft_time on step == 0 always more slower than normal_time (need to find the reason)

The reason is the draft_kv_cache convert from bf16 (malloc new memory) will call extend_kv_cache when  draft_step== 0, So we can init the needed kv_cache and reuse it each iter.

With the same chinest promt:
spr01 | Llama-2-13b-chat-hf | chatglm3-6b| Qwen-14B
-- | -- | -- | -- |
before draft_time | 74.752 | 34.252 | 61.053
before target_time | 200.961 | 101.003| 186.477
after draft_time | 64.221 | 29.139 | 53.270
after target_time | 201.692 | 102.653| 188.057 |

### 4. How to test?
- [ ] N/A
- [ ] Unit test
- [x] Application test
- [ ] Document test
- [ ] ...

### 5. New dependencies

- [ ] New Python dependencies
       - Dependency1 
       - Dependency2
       - ...
- [ ] New Java/Scala dependencies and their license
       - Dependency1 and license1
       - Dependency2 and license2
       - ...

### CODE REVIEW

File: python/llm/src/bigdl/llm/transformers/speculative.py [hunk 1]
```diff
@@ -195,6 +195,10 @@ def speculative_generate(self,
             # Draft model auto-regressively generate k tokens
             # Early stop when prob less then th_stop_draft
             for step_draft in range(max_step_draft):
+                if self.device.type == 'cpu':
+                    past_key_values = [
```
Comment (Uxito-Ada) [inline python/llm/src/bigdl/llm/transformers/speculative.py:199]: Is this repeated every step, which may cause reductant latency and memory copy?

Comment (Uxito-Ada) [reply python/llm/src/bigdl/llm/transformers/speculative.py:199]: could it be done once e.g. when init, and only process the additive part in the following?

Comment (jason-dai) [reply python/llm/src/bigdl/llm/transformers/speculative.py:199]: > could it be done once e.g. when init, and only process the additive part in the following?

Yes, we should try that.

Comment (jason-dai) [reply python/llm/src/bigdl/llm/transformers/speculative.py:199]: Is this implemented in the same when during our previous testing?

**Review by glorysdj — APPROVED** [review]
LGTM

File: python/llm/src/bigdl/llm/transformers/speculative.py [hunk 2]
```diff
@@ -189,7 +190,71 @@ def speculative_generate(self,
...
+                                        dtype=torch.float32)
+                        v0 = torch.ones(len0, len1, len2 + max_new_tokens, len3,
+                                        dtype=torch.float32)
+                        past_key_values1.append((k0, v0))
+                        past_key_values1[i][0][:, :, :len2, :] = past_key_values[i][0].to(
+                            torch.float32)
+                        past_key_values1[i][1][:, :, :len2, :] = past_key_values[i][1].to(
+                            torch.float32)
+
+            # each iter cut off cur_len kv_cache from past_key_values1
```
Comment (cyita) [inline python/llm/src/bigdl/llm/transformers/speculative.py:236]: Maybe move these two lines to `if self.device.type == 'cpu':`?

**Comment by cyita** [comment]
This PR won't affect the GPU performance.

**Pull request merged**